Remove processing acks that are older than the one client loaded from

agarwal-navin opened the pull request:
Currently, when a summarizer client first loads, it will most likely process the ack of the summary it loaded from. This is totally unnecessary as this ack gets ignored. We can see this in telemetry today as shown below where `refreshLatestSummary` logs are soon as the Summarizer starts before it even attempts a summary:
![image](https://github.com/microsoft/FluidFramework/assets/45832642/04bf079d-ef32-47eb-919a-0952094747ef)

A summarizer client should only care about acks for summaries that it generated itself. Summarizer client connects in "write" mode and it waits for connected state before it creates `RunningSummarizer`. So, by the time `RunningSummarizer` is created, all older acks will have been processed.

This change removes the logic that processes the last ack seen when `RunningSummarizer` starts. When it starts waiting for acks in a loop, it starts waiting for acks from the summary next to the one it loaded from. This will ensure that summarizer client never processes ack for the summary it loaded from.

Note: It's possible in rare cases that while it was catching up, it saw an ack that was newer than the one it loaded from and was from another summarizer running in parallel. In this scenario, the summarizer closes today during ack processesing flow because it never processes newer acks from other summarizer clients. With this change, in this scenario, the summarizer will instead close when uploading the summary because its parent will be incorrect.

agarwal-navin committed:
Remove processing acks that are older than the one client loaded from

agarwal-navin committed:
Move

agarwal-navin commented on packages/runtime/container-runtime/src/summary/runningSummarizer.ts:
```diff
@@ -272,94 +268,73 @@ export class RunningSummarizer extends TypedEventEmitter<ISummarizerEvents> impl
-			// we record undefined, not previous handles.
-			await this.summaryCollection.waitSummaryAck(refSequenceNumber);
-
-			summaryLogger.sendTelemetryEvent({
```
Removed this log because it doesn't add any value. Container runtime already logs in this flow and that's the one we look at and has more information.

rajatch-ff replied:
And there's no point of failure between this log location and container runtime log location? If there was a point of failure, this log could prove useful in debugging scenarios where the code-flow was successful till this point, but failed later on.

agarwal-navin replied:
No, there isn't.

rajatch-ff commented on packages/runtime/container-runtime/src/summary/runningSummarizer.ts:
```diff
@@ -272,94 +268,73 @@ export class RunningSummarizer extends TypedEventEmitter<ISummarizerEvents> impl
-			lastAckRefSeq > 0 ? lastAckRefSeq : this.runtime.deltaManager.initialSequenceNumber;
+	private async processIncomingSummaryAcks() {
+		// Start waiting for acks that are for summaries newer that the one this client loaded from.
+		let referenceSequenceNumber = this.runtime.deltaManager.initialSequenceNumber + 1;
```
>referenceSequenceNumber [](http://example.com/codeflow?start=6&length=23)

nit: `nextRefSequenceNumber` would resonate more with the semantic meaning here and on line 337.

agarwal-navin replied:
Good point. Renamed it.

rajatch5 reviewed (approved):
Looks good. Thanks for the good description of the issue and the fix. Helped a ton in setting context. Suggestions I left are non-blocking and optional.

agarwal-navin committed:
PR comments 1

agarwal-navin committed:
Fix tests

agarwal-navin commented on packages/test/test-end-to-end-tests/src/test/gc/gcStateResetInSummaries.spec.ts:
```diff
@@ -306,87 +306,6 @@ describeCompat("GC state reset in summaries", "NoCompat", (getTestObjectProvider
-	 * enabled and state is refreshed from a snapshot that had GC disabled, the GC state needs to be reset and a full
-	 * tree summary should happen.
-	 */
-	it("regenerates GC state on refreshing from snapshot that has GC disabled", async () => {
```
Removed this test because it does not test anything unique. This scenario is already tested in `gcLoadingFromOlderSummaries.spec.ts`

tyler-cai-microsoft replied:
I like these changes

agarwal-navin commented on packages/test/test-end-to-end-tests/src/test/gc/gcLoadingFromOlderSummaries.spec.ts:
```diff
@@ -143,17 +143,20 @@ describeCompat("GC loading from older summaries", "NoCompat", (getTestObjectProv
 		await summarizer2.summarizeOnDemand({ reason: "test" }).summarySubmitted;
+
+		// Wait for the container for the above summarizer to be disposed.
+		await timeoutPromise((resolve) => {
```
This test needed update because previously, the ack from another summarizer was processed when the running summarizer started and before summarizeOnDemand was called. Now, the ack is processed async and there is no guarantee that it will be processed by the time summarizeOnDemand is called.

tyler-cai-microsoft replied:
All we have to do is make sure the ack is processed right? Isn't there a way to ensure we've processed s1's ack via just listening to the container/containerRuntime's ops?

agarwal-navin replied:
The problem is that just observing the ack isn't enough since the processing happens asynchronously. There has to be some trigger after the ack processing is complete and we don't have any.

tyler-cai-microsoft replied:
Feel free to resolve, I just ran into this issue myself

markfields commented:
> It's possible in rare cases that while it was catching up, it saw an ack that was newer than the one it loaded from and was from another summarizer running in parallel

Can we revisit the data we have in 1P telemetry to confirm this is as rare as we expect?

markfields commented on packages/runtime/container-runtime/src/summary/runningSummarizer.ts:
```diff
@@ -272,94 +268,74 @@ export class RunningSummarizer extends TypedEventEmitter<ISummarizerEvents> impl
-				summaryLogger.sendTelemetryEvent({
-					eventName: "RefreshAttemptWithSummarizerRunning",
-					referenceSequenceNumber: refSequenceNumber,
+	private async handleSummaryAck(lastAck: IAckedSummary) {
```
nit: why not just call it `ack`?

markfields replied:
Also let's add a comment to this function

markfields commented on packages/test/test-end-to-end-tests/src/test/gc/gcLoadingFromOlderSummaries.spec.ts:
```diff
@@ -143,17 +143,20 @@ describeCompat("GC loading from older summaries", "NoCompat", (getTestObjectProv
+			if (container2.disposed) {
+				resolve();
+			}
+			container2.on("disposed", () => {
+				resolve();
+			});
```
nit: you can just pass `resolve` as the 2nd arg
```suggestion
			container2.on("disposed", resolve);
```

markfields commented on packages/test/test-end-to-end-tests/src/test/gc/gcLoadingFromOlderSummaries.spec.ts:
```diff
@@ -143,17 +143,20 @@ describeCompat("GC loading from older summaries", "NoCompat", (getTestObjectProv
+		// Wait for the container for the above summarizer to be disposed.
+		await timeoutPromise((resolve) => {
+			if (container2.disposed) {
+				resolve();
```
Might as well return here to avoid the unused event below.

markfields replied:
Not sure how careful we have to be about leaking listeners in mocha tests. If we need to be careful, then you should do the thing where the listener calls `.off` and passes itself.  Or just use `.once` I guess

agarwal-navin replied:
I thought resolve will do that.

agarwal-navin committed:
PR comments

tyler-cai-microsoft commented on packages/runtime/container-runtime/src/summary/runningSummarizer.ts:
```diff
@@ -91,21 +91,17 @@ export class RunningSummarizer extends TypedEventEmitter<ISummarizerEvents> impl
 			);
 		});
 
+		await summarizer.waitStart();
```
What do you think the impact of moving wait start to below processing incoming acks will do? Will it have a restart impact?

tyler-cai-microsoft replied:
Do you know what `waitStart` does? It doesn't immediately make sense to me when I read through the code

agarwal-navin replied:
Yeah, `waitStart` will wait for a summary ack if it sees a summary op that hasn't been acked. The idea being that if the previous summarizer closed before it saw an ack, this summarizer would do it.

It may need to happen before we summarize so that the summary doesn't fail. There are 3 cases where we see an op but not an ack:
- The previous summary succeeded but we loaded from the summary before that. In this case, if we don't process the ack, then the current summary will fail.
- The previous summary succeeded and we loaded from it. In this case, we don't need to process this ack and this change will ensure that.
- The previous summary failed. In this case, waiting for the ack just adds a delay but there is no way to know it.

agarwal-navin replied:
As for the impact, the kicking of summary heuristics and processing ack are both async and they race. Before this change, if there was a scenario like case # 1 above, chances are that we wouid start a summarization before processing the ack and the summary would fail. Now, we will process the ack first and that will fail which is cheaper and faster that doing the summary and failling.

agarwal-navin replied:
I actually reverted back the order. Also, added back synchronously processing the first ack but only if its newer that the one the client loaded from. Added a comment to explain why it is needed.

tyler-cai-microsoft replied:
Just took a look, looks good

tyler-cai-microsoft reviewed (approved):
Looks good to me

agarwal-navin committed:
Revert processing the first ack

agarwal-navin commented:
> > It's possible in rare cases that while it was catching up, it saw an ack that was newer than the one it loaded from and was from another summarizer running in parallel
> 
> Can we revisit the data we have in 1P telemetry to confirm this is as rare as we expect?

@markfields Good callout. I looked at the data and there were a few instances of this happening. Digging a little deeper made me realize that this happens when summarizer loads from a cached snapshot and the previous summarizer (same client probably) did a successful summary.
So, I added back synchronously processing the first ack but only if its newer that the one the client loaded from. That will hep with this scenario. Added a comment to explain why it is needed.
